README.md
This repo draws heavily on the code provided by Seeed Studio and utilizes their MLX90640 110 degree Thermal Camera.

The frame rate setting for the camera in this code was chosen based on initial experiences with a Raspberry Pi Zero W.

The camera uses a 32x24 array but the data feed is all one line.

# Setup
## Wiring

The Seeed Studio Grove Thermal Imaging Camera IR Array uses 3 to 6V. I connected mine to the 3.3V DC
on my Raspberry Pi Zero W. Current consumption of approximately 18mA.

Match the *SDA* and *SCL* pins on the camera to those on the Pi. 

- Camera SDA -> GPIO 2 (SDA)
- Camera SCL -> GPIO 3 (SCL)

## Software

1. Install `grove.py`.
    ```shell script
    curl -sL https://github.com/Seeed-Studio/grove.py/raw/master/install.sh | sudo bash -s -
    ```
1. Install the driver. 
    ```shell script
    pip3 install seeed-python-mlx90640
   ```
   If you need to run with sudo.
   ```shell script
    sudo pip3 install seeed-python-mlx90640
    ```
1. Check for i2c number. This should return a device.
    ```shell script
    ls /dev/i2c*
    ```
1. Check for the camera as an i2c device. Its address is `0x33`.
    ```shell script
    i2cdetect -y -r 1
    ```
1. Give it a try! :D

### Refresh Rate

I found that `REFRESH_2_HZ` worked on a Raspberry Pi Zero W with `stream.py`. Higher ones might work, especially for
`time-lapse.py`.

This code is an example of how to create the device and set the refresh rate. This was found 
on [Seed Studio Wiki for their Thermal Imaging Camera](http://wiki.seeedstudio.com/Grove-Thermal-Imaging-Camera-IR-Array/). 

```python
import seeed_mlx90640
mlx = seeed_mlx90640.grove_mxl90640()
mlx.refresh_rate = seeed_mlx90640.RefreshRate.REFRESH_8_HZ  # The fastest for raspberry 4 
# REFRESH_0_5_HZ = 0b000  # 0.5Hz
# REFRESH_1_HZ = 0b001  # 1Hz
# REFRESH_2_HZ = 0b010  # 2Hz
# REFRESH_4_HZ = 0b011  # 4Hz
# REFRESH_8_HZ = 0b100  # 8Hz
# REFRESH_16_HZ = 0b101  # 16Hz
# REFRESH_32_HZ = 0b110  # 32Hz
# REFRESH_64_HZ = 0b111  # 64Hz
```

#### Boosting Refresh Rate

Always be aware when editing boot options your Pi might not boot.

Find the `dtparam=i2c_arm=on` line in `/boot/config.txt` and replace it with the one below.

```shell script
dtparam=i2c_arm=on,i2c_arm_baudrate=400000
```

This sets the i2c baudrate to the maximum supported by the CPU, the default is `100000`. See the links below for 
peripheral specification and device tree.

# Streaming

Run the `stream.py` script to host a small website with Flask that streams what the thermal camera sees.

# Thermal Image Capturing

Run `time-lapse.py` or write a simple script to call `thermal.py` to capture the thermal data.

# Time Lapse Images

Run `time-lapse.py`. Data is written to `thermal.csv`. Images are captured as raw numeric data from the camera 
every 10 seconds and stored in the CSV file. Use `thermal-to-images.py` to convert the raw data into PNG images.

The data is formatted as one line per image. An example file is `thermal.csv`.

## Converting Images to Video

I found that I needed an image that is `images/img0000.png` for this to work. I don't know if everything has to

```shell script
ffmpeg -r 5 -i images/img%04d.png -vcodec libx264 -pix_fmt yuv420p -y movie.mp4
```

# Sources

* [Seeed Studio Thermal Camera Wiki](http://wiki.seeedstudio.com/Grove-Thermal-Imaging-Camera-IR-Array/)
* [Seeed Studio Thermal Camera Product Page](https://www.seeedstudio.com/Grove-Thermal-Imaging-Camera-IR-Array-MLX90640-110-degree-p-4334.html)
* [Video Streaming with Flask Source Code](https://github.com/miguelgrinberg/flask-video-streaming)
* [Video Streaming with Flask](https://blog.miguelgrinberg.com/post/video-streaming-with-flask)
  * `Video Streaming with Flask.pdf`
* [Flask Video Streaming Revisited](https://blog.miguelgrinberg.com/post/flask-video-streaming-revisited) 
  * `Flask Video Streaming Revisited.pdf`
* [Raspberry Pi BCM2835 Peripheral Specification](https://www.raspberrypi.org/documentation/hardware/raspberrypi/bcm2835/README.md)
* [Raspberry Pi Device Tree Overlays](https://github.com/raspberrypi/firmware/blob/master/boot/overlays/README)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2019-01-30
### Added
- Notes on increasing the i2c baudrate.
- `__pycache__` to `.gitignore`.

### Changed
- Changed `time-lapse.py` so it runs faster (i2c baudrate increased on Pi), has currently commented out lines
  for limiting how many images it captures and moved some lines of code around to try and boost performance.

### Removed
- Image range sub-set from `thermal-to-images.py`.
  
### Fixed
- Added exception catch to `thermal-to-images.py` for a `ValueError`. Caused by `Nan` conversion to float.

## [1.0.1] - 2020-01-29
### Added
- Blank line at the end of `requirements.txt` for clarity.

### Fixed
- `camera.py` was loading removed images, this was no longer needed.
- `camera.py` was calling the wrong method.
- `thermal.py` was not calling the class method.

## [1.0.0] - 2020-01-28
### Added
- Initial version.
- `README.md` has details on the repo.


view.py
import colorsys
import csv

from PIL import Image


def mapValue(value, curMin, curMax, desMin, desMax):
    curDistance = value - curMax
    if curDistance == 0:
        return desMax
    curRange = curMax - curMin
    direction = 1 if curDistance > 0 else -1
    ratio = curRange / curDistance
    desRange = desMax - desMin
    value = desMax + (desRange / ratio)
    return value


def constrain(value, down, up):
    value = up if value > up else value
    value = down if value < down else value
    return value


def isDigital(value):
    try:
        if value == "nan":
            return False
        else:
            float(value)
        return True
    except ValueError:
        return False


def process_data(hetData):
    maxHet = 0
    minHet = 500
    tempData = []
    nanCount = 0

    if len(hetData) < 768:
        return

    for i in range(0, 768):
        curCol = i % 32
        newValueForNanPoint = 0
        curData = None

        if i < len(hetData) and isDigital(hetData[i]):
            curData = float(hetData[i])
        else:
            interpolationPointCount = 0
            sumValue = 0
            # print("curCol",curCol,"i",i)

            abovePointIndex = i - 32
            if (abovePointIndex > 0):
                if hetData[abovePointIndex] is not "nan":
                    interpolationPointCount += 1
                    sumValue += float(hetData[abovePointIndex])

            belowPointIndex = i + 32
            if (belowPointIndex < 768):
                print(" ")
                if hetData[belowPointIndex] is not "nan":
                    interpolationPointCount += 1
                    sumValue += float(hetData[belowPointIndex])

            leftPointIndex = i - 1
            if (curCol != 31):
                if hetData[leftPointIndex] is not "nan":
                    interpolationPointCount += 1
                    sumValue += float(hetData[leftPointIndex])

            rightPointIndex = i + 1
            if (belowPointIndex < 768):
                if (curCol != 0):
                    if hetData[rightPointIndex] is not "nan":
                        interpolationPointCount += 1
                        sumValue += float(hetData[rightPointIndex])

            curData = sumValue / interpolationPointCount
            # For debug :
            # print(abovePointIndex,belowPointIndex,leftPointIndex,rightPointIndex)
            # print("newValueForNanPoint",newValueForNanPoint," interpolationPointCount" , interpolationPointCount ,"sumValue",sumValue)
            nanCount += 1

        tempData.append(curData)
        maxHet = tempData[i] if tempData[i] > maxHet else maxHet
        minHet = tempData[i] if tempData[i] < minHet else minHet

    if maxHet == 0 or minHet == 500:
        return
    # For debug :
    # if nanCount > 0 :
    #     print("____@@@@@@@ nanCount " ,nanCount , " @@@@@@@____")

    return {
        "frame": tempData,
        "maxHet": maxHet,
        "minHet": minHet
    }


with open('thermal2.csv') as csvfile:
    data = list(csv.reader(csvfile))[0]  # 768 values

# Make some RGB values.
# # Cycle through hue vertically & saturation horizontally
# colors = []
# for i in data:
#     # Convert color from HSV to RGB
#     rgb = colorsys.hsv_to_rgb(float(i) / 360, 1, 1)
#     rgb = [int(0.5 + 255 * u) for u in rgb]
#     colors.extend(rgb)
#
# # Convert list to bytes
# colors = bytes(colors)
# img = Image.frombytes('RGB', (32, 24), colors)
# img.show()
# img.save('hues.png')

frame = process_data(data)

maxHet = frame["maxHet"]
minHet = frame["minHet"]
frame = frame["frame"]

pixelSize = 15
width = 480
height = 360
minHue = 180
maxHue = 360

# tempData = constrain(mapValue(frame[index], minHet, maxHet, minHue, maxHue), minHue, maxHue)

colors = []
for i in frame:
    # Convert color from HSV to RGB
    tempData = constrain(mapValue(i, minHet, maxHet, minHue, maxHue), minHue, maxHue)
    rgb = colorsys.hsv_to_rgb(tempData / 360, 1.0, 1.0)
    rgb = [int(0.5 + 255 * u) for u in rgb]
    colors.extend(rgb)

# Convert list to bytes
colors = bytes(colors)
img = Image.frombytes('RGB', (32, 24), colors)
img = img.rotate(90)
img = img.resize((width, height))
img.show()
img.save('hues2.png')

stream.py
"""Run this script to host a simple web page that streams the thermal camera images."""

from flask import Flask, render_template, Response

from camera import Camera

app = Flask(__name__)


@app.route('/')
def index():
    """Video streaming home page."""
    return render_template('index.html')


def gen(camera):
    """Video streaming generator function."""
    while True:
        frame = camera.get_frame()
        yield (b'--frame\r\n'
               b'Content-Type: image/jpeg\r\n\r\n' + frame + b'\r\n')


@app.route('/video_feed')
def video_feed():
    """Video streaming route. Put this in the src attribute of an img tag."""
    return Response(gen(Camera()),
                    mimetype='multipart/x-mixed-replace; boundary=frame')


if __name__ == '__main__':
    app.run(host='0.0.0.0', threaded=True)  # Defaults to port 5000


thermal.py
"""Read data from the thermal camera and return a bytes object with the image data."""

import io

import seeed_mlx90640

from thermal_data import ThermalData


class Thermal:

    def __init__(self):
        self.mlx = seeed_mlx90640.grove_mxl90640()
        self.mlx.refresh_rate = seeed_mlx90640.RefreshRate.REFRESH_2_HZ  # The fastest for Raspberry Pi Zero W

        self.thermal_data = ThermalData()

    def get_serial_data(self):
        """

        :return: bytes object containing image data.
        """

        data = [0] * 768

        while True:
            try:
                self.mlx.getFrame(data)
            except ValueError:
                continue
            break

        img = self.thermal_data.get_data(data)
        imgByteArr = io.BytesIO()
        img.save(imgByteArr, format='PNG')
        imgByteArr = imgByteArr.getvalue()

        return imgByteArr


thermal_data.py
"""Image data processing for the thermal camera.

The methods below were copied from https://github.com/gobuyun/seeed_ircamera/blob/master/seeed_python_ircamera.py. A
copy of this code can be found in this repo in the file: `seeed_python_ircamera.py`.
"""

import colorsys

from PIL import Image


class ThermalData:

    def __init__(self):
        pass

    @staticmethod
    def mapValue(value, curMin, curMax, desMin, desMax):
        curDistance = value - curMax
        if curDistance == 0:
            return desMax
        curRange = curMax - curMin
        direction = 1 if curDistance > 0 else -1
        ratio = curRange / curDistance
        desRange = desMax - desMin
        value = desMax + (desRange / ratio)
        return value

    @staticmethod
    def constrain(value, down, up):
        value = up if value > up else value
        value = down if value < down else value
        return value

    @staticmethod
    def isDigital(value):
        try:
            if value == "nan":
                return False
            else:
                float(value)
            return True
        except ValueError:
            return False

    def process_data(self, hetData):
        maxHet = 0
        minHet = 500
        tempData = []
        nanCount = 0

        if len(hetData) < 768:
            return

        for i in range(0, 768):
            curCol = i % 32
            newValueForNanPoint = 0
            curData = None

            if i < len(hetData) and self.isDigital(hetData[i]):
                curData = float(hetData[i])
            else:
                interpolationPointCount = 0
                sumValue = 0
                # print("curCol",curCol,"i",i)

                abovePointIndex = i - 32
                if (abovePointIndex > 0):
                    if hetData[abovePointIndex] is not "nan":
                        interpolationPointCount += 1
                        sumValue += float(hetData[abovePointIndex])

                belowPointIndex = i + 32
                if (belowPointIndex < 768):
                    print(" ")
                    if hetData[belowPointIndex] is not "nan":
                        interpolationPointCount += 1
                        sumValue += float(hetData[belowPointIndex])

                leftPointIndex = i - 1
                if (curCol != 31):
                    if hetData[leftPointIndex] is not "nan":
                        interpolationPointCount += 1
                        sumValue += float(hetData[leftPointIndex])

                rightPointIndex = i + 1
                if (belowPointIndex < 768):
                    if (curCol != 0):
                        if hetData[rightPointIndex] is not "nan":
                            interpolationPointCount += 1
                            sumValue += float(hetData[rightPointIndex])

                curData = sumValue / interpolationPointCount
                # For debug :
                # print(abovePointIndex,belowPointIndex,leftPointIndex,rightPointIndex)
                # print("newValueForNanPoint",newValueForNanPoint," interpolationPointCount" , interpolationPointCount ,"sumValue",sumValue)
                nanCount += 1

            tempData.append(curData)
            maxHet = tempData[i] if tempData[i] > maxHet else maxHet
            minHet = tempData[i] if tempData[i] < minHet else minHet

        if maxHet == 0 or minHet == 500:
            return
        # For debug :
        # if nanCount > 0 :
        #     print("____@@@@@@@ nanCount " ,nanCount , " @@@@@@@____")

        return {
            "frame": tempData,
            "maxHet": maxHet,
            "minHet": minHet
        }

    def get_data(self, data):

        # with open('thermal2.csv') as csvfile:
        #     data = list(csv.reader(csvfile))[0]  # 768 values

        # Make some RGB values.
        # # Cycle through hue vertically & saturation horizontally
        # colors = []
        # for i in data:
        #     # Convert color from HSV to RGB
        #     rgb = colorsys.hsv_to_rgb(float(i) / 360, 1, 1)
        #     rgb = [int(0.5 + 255 * u) for u in rgb]
        #     colors.extend(rgb)
        #
        # # Convert list to bytes
        # colors = bytes(colors)
        # img = Image.frombytes('RGB', (32, 24), colors)
        # img.show()
        # img.save('hues.png')

        frame = self.process_data(data)

        maxHet = frame["maxHet"]
        minHet = frame["minHet"]
        frame = frame["frame"]

        pixelSize = 15
        width = 480
        height = 360
        minHue = 180
        maxHue = 360

        # tempData = constrain(mapValue(frame[index], minHet, maxHet, minHue, maxHue), minHue, maxHue)

        colors = []
        for i in frame:
            # Convert color from HSV to RGB
            tempData = self.constrain(self.mapValue(i, minHet, maxHet, minHue, maxHue), minHue, maxHue)
            rgb = colorsys.hsv_to_rgb(tempData / 360, 1.0, 1.0)
            rgb = [int(0.5 + 255 * u) for u in rgb]
            colors.extend(rgb)

        # Convert list to bytes
        colors = bytes(colors)
        img = Image.frombytes('RGB', (32, 24), colors)
        img = img.rotate(90)
        img = img.transpose(Image.FLIP_LEFT_RIGHT)
        img = img.resize((width, height))

        return img
        # img.show()
        # img.save('hues2.png')
